drivers/dht/dht.c

```diff
@@ -19,192 +20,318 @@
  *
  * @}
  */
 
+#include <errno.h>
 #include <stdint.h>
 #include <string.h>
 
-#include "log.h"
 #include "assert.h"
-#include "xtimer.h"
-#include "timex.h"
-#include "periph/gpio.h"
-
 #include "dht.h"
 #include "dht_params.h"
+#include "log.h"
+#include "periph/gpio.h"
+#include "time_units.h"
+#include "ztimer.h"
+
+#if __AVR__
+/* On AVR, `gpio_read()` takes 2 µs mainly due to the port register
+ * 16-bit address computation, so we will need those libs to compute it
+ * once and for all. */
+#include <atmega_pcint.h>
+#include "atmega_gpio.h"
+/* On AVR, `ztimer_spin()` is not enough accurate on delays less than 100 µs
+ * (~40 µs on a 16 MHz atmega328p for `ztimer_spin(ZTIMER_USEC, 1)`) so
+ * we will use the avr-libc `_delay_us()` function instead. */
+#define F_CPU                   (CLOCK_CORECLOCK)
+#define __DELAY_BACKWARD_COMPATIBLE__
+#include "util/delay.h"
+#endif
 
 #define ENABLE_DEBUG            0
 #include "debug.h"
 
 /* Every pulse send by the DHT longer than 40µs is interpreted as 1 */
-#define PULSE_WIDTH_THRESHOLD   (40U)
-/* If an expected pulse is not detected within 1000µs, something is wrong */
-#define TIMEOUT                 (1000U)
-/* The DHT sensor cannot measure more than once a second */
-#define DATA_HOLD_TIME          (US_PER_SEC)
-/* The start signal by pulling data low for at least 18ms and then up for
- * 20-40µs*/
-#define START_LOW_TIME          (20U * US_PER_MS)
-#define START_HIGH_TIME         (40U)
-
-static inline void _reset(dht_t *dev)
+#define READ_THRESHOLD   (40U)
+/* If an expected pulse is not detected within 85µs, something is wrong */
+#define SPIN_TIMEOUT            (85U)
+/* According to datasheet the DHT sensor cannot measure more than once
+ * every 2 seconds but most are faster */
+#define RESPAWN_TIMEOUT         (2U * US_PER_SEC)
+/* The start signal by pulling data low for at least 18 ms for DHT11, at
+ * most 20 ms (AM2301 / DHT22 / DHT21). Then release the bus and the
+ * sensor should respond by pulling data low for 80 µs, then release for
+ * 80µs before start sending data. */
+#define START_LOW_TIME          (19U * US_PER_MS)
+#define START_THRESHOLD         (40U)
+/* DHTs have to wait for power 1 or 2 seconds depending on the model */
+#define POWER_WAIT_TIMEOUT      (2U * US_PER_SEC)
+
+enum {
+    BYTEPOS_HUMIDITY_HIGH = 0,
+    BYTEPOS_HUMIDITY_LOW = 1,
+    BYTEPOS_TEMPERATURE_HIGH = 2,
+    BYTEPOS_TEMPERATURE_LOW = 3,
+    BYTEPOS_CHECKSUM = 4,
+};
+
+struct dht_data {
+#ifdef __AVR__
+    uint16_t p_in_addr;
+    uint8_t p_in_mask;
+#else
+    gpio_t pin;
+#endif
+    gpio_mode_t in_mode;
+    uint8_t data[5];
+    int8_t bit_pos;
+    uint8_t bit;
+};
+
+#ifdef __AVR__
+static inline void _wait_for_level(uint16_t p_in_addr,
+                                    uint8_t p_in_mask, bool expected)
+{
+    uint8_t timeout = SPIN_TIMEOUT;
+    while (((_SFR_MEM8(p_in_addr) & p_in_mask) != expected) && timeout--) {
+        _delay_us(1);
```

Comment:
Maybe `delay_us_spinning()` from https://github.com/RIOT-OS/RIOT/pull/19719 could solve this? It adds an implementation not based on ztimer for AVR that is pretty much identical to the one from the AVR libc, while still using `ztimer_spin(ZTIMER_USEC, <num>)` on other platforms.

Reply:
The issue with ztimer_spin() was that it called ztimer_now() twice, which is the actual time-consuming function. I removed all calls to ztimer_spin() and managed to make it work by minimizing the usage of ztimer_now() as much as possible.
Without questioning the usefulness of the sys/wait API, I believe that wait_us_spinning() creates a surprising mix of kind. On one hand, we have three functions that prioritize efficiency over accuracy, and on the other hand, we have a function that does the complete opposite! Ultimately, I wonder if a delay function really has its place in a multitasking operating system...